storage: Update timestamp cache before updating a lease on a replica

m-schneider opened the pull request:
- Before we would update the in-memory lease on a replica,
unlock it and then we would update the timestamp cache. This would
allow a replica to start processing requests for a lease without
first having taken into account reads served from the old lease
holder. We could see this by running:
make stress PKG=./pkg/sql/logictest
TESTS=TestParallel/subquery_retry_multinode STRESSFLAGS='-p 32'

This would fail because we would have duplicate values in a table
where inserts should have had monotonically increasing values.

- After this change the test passes. We also updated splitPostApply,
to send the current lease as the previous lease to leasePostApply to
prevent an update to the timestamp cache when a lease hasn't changed
hands.

Closes #16610

Commit:
storage: Update timestamp cache before updating a lease on a replica

- Before we would update the in-memory lease on a replica,
unlock it and then we would update the timestamp cache. This would
allow a replica to start processing requests for a lease without
first having taken into account reads served from the old lease
holder. We could see this by running:
make stress PKG=./pkg/sql/logictest
TESTS=TestParallel/subquery_retry_multinode STRESSFLAGS='-p 32'

This would fail because we would have duplicate values in a table
where inserts should have had monotonically increasing values.

- After this change the test passes. We also updated splitPostApply,
to send the current lease as the previous lease to leasePostApply to
prevent an update to the timestamp cache when a lease hasn't changed
hands.

Closes #16610

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/17109)
<!-- Reviewable:end -->

tschottdorf commented:
Looks good so far!

---

Reviewed 2 of 2 files at r1.
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/storage/store.go, line 1861 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17109#-KpR9wuBDvXuQ-8ExBe1:-KpR9vOfAN3QvIz2Sn2-:bhqkow7) ([raw file](https://github.com/cockroachdb/cockroach/blob/05e186149b262a5d7838129d64cb62caf03067dc/pkg/storage/store.go#L1861)):*
> ```Go
> 	// the PushTxnQueue. Note that we pass in an right lease for prevLease so
> 	// that we don't unnecessarily update the timestamp cache.
> 	rightRng.leasePostApply(ctx, rightLease, rightReplicaID, rightLease)
> ```

Can you just drop the last argument to `leasePostApply` and instead load `prevLease` in `leasePostApply`? In this callsite, we already have `rightRng.Lease == rightLease` so it will continue to be a noop, and similarly the other call site won't have to fetch the lease manually. I think the same holds for the `ReplicaID` which in both cases is picked from `mu` which you may as well do inside the function.

---

*[pkg/storage/store.go, line 1861 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17109#-KpRByMYvUxbiOYNZ9sP:-KpRByMYvUxbiOYNZ9sQ:b-6a5lt6) ([raw file](https://github.com/cockroachdb/cockroach/blob/05e186149b262a5d7838129d64cb62caf03067dc/pkg/storage/store.go#L1861)):*
> ```Go
> 	// the PushTxnQueue. Note that we pass in an right lease for prevLease so
> 	// that we don't unnecessarily update the timestamp cache.
> 	rightRng.leasePostApply(ctx, rightLease, rightReplicaID, rightLease)
> ```

Looking through the `git blame` for this call to `leasePostApply`, I found this: https://github.com/cockroachdb/cockroach/pull/14305

It seems that the original call intentionally wanted things to look like the lease was changing hands so that `pushTxnQueue.Clear()` is called in `leasePostApply`. ISTM that we've changed this back now; is the test that was added in that PR not failing? Can you check what's going on there?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17109#-:-KpRBtGRcu9aYatZpbQ0:b-37o3tw)*
<!-- Sent from Reviewable.io -->

m-schneider commented:
Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions.

---

*[pkg/storage/store.go, line 1861 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17109#-KpRByMYvUxbiOYNZ9sP:-KpRcrbbWOgSLLMIHaj9:b37783y) ([raw file](https://github.com/cockroachdb/cockroach/blob/05e186149b262a5d7838129d64cb62caf03067dc/pkg/storage/store.go#L1861)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Looking through the `git blame` for this call to `leasePostApply`, I found this: https://github.com/cockroachdb/cockroach/pull/14305

It seems that the original call intentionally wanted things to look like the lease was changing hands so that `pushTxnQueue.Clear()` is called in `leasePostApply`. ISTM that we've changed this back now; is the test that was added in that PR not failing? Can you check what's going on there?
</blockquote></details>

I reran the storage tests and nothing fails. I read through all of the comments for that PR as well and the main concern was for the pushTxnQueue to be enabled after the call to leasePostApply, which it is because that's only gated by iAmTheLeaseHolder.  The tests added in that PR only check to see if the queue is enabled, but don't test the contents.

---

*[pkg/storage/store.go, line 1861 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17109#-KpR9wuBDvXuQ-8ExBe1:-KpRd5Bqr2851FgOVRG3:b-4sjgjv) ([raw file](https://github.com/cockroachdb/cockroach/blob/05e186149b262a5d7838129d64cb62caf03067dc/pkg/storage/store.go#L1861)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Can you just drop the last argument to `leasePostApply` and instead load `prevLease` in `leasePostApply`? In this callsite, we already have `rightRng.Lease == rightLease` so it will continue to be a noop, and similarly the other call site won't have to fetch the lease manually. I think the same holds for the `ReplicaID` which in both cases is picked from `mu` which you may as well do inside the function.
</blockquote></details>

Good point, cleaned up.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17109)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[pkg/storage/store.go, line 1861 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17109#-KpRByMYvUxbiOYNZ9sP:-KpVNsXjT5PTu-6JERyl:b-4tr286) ([raw file](https://github.com/cockroachdb/cockroach/blob/05e186149b262a5d7838129d64cb62caf03067dc/pkg/storage/store.go#L1861)):*
<details><summary><i>Previously, m-schneider (Masha Schneider) wrote…</i></summary><blockquote>

I reran the storage tests and nothing fails. I read through all of the comments for that PR as well and the main concern was for the pushTxnQueue to be enabled after the call to leasePostApply, which it is because that's only gated by iAmTheLeaseHolder.  The tests added in that PR only check to see if the queue is enabled, but don't test the contents.
</blockquote></details>

Oh, gotcha, I was looking at the wrong part of the code, namely the one where the queue is disabled. No problem then!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17109#-:-KpVOXkToZp37Esnsdi6:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
See https://teamcity.cockroachdb.com/viewLog.html?buildId=300601&buildTypeId=Cockroach_UnitTests for fixing the CI.

m-schneider commented:
Thanks!

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17109#-:-KpV_Z9Nwj--5jYgg2an:b-ucfknh)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
I think we want to cherry-pick this into 1.0.x. If we do it soon it can go in to 1.0.4 (the cutoff is today).

---

Review status: 1 of 2 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/storage/store.go, line 1858 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/17109#-KpVcexgpGpYOJK8V5KO:-KpVcexgpGpYOJK8V5KP:b7zix7q) ([raw file](https://github.com/cockroachdb/cockroach/blob/c54d9b059b482ac965ccbeb083d412d98a862200/pkg/storage/store.go#L1858)):*
> ```Go
> 	// Invoke the leasePostApply method to ensure we properly initialize
> 	// the replica according to whether it holds the lease. This enables
> 	// the PushTxnQueue. Note that we pass in an right lease for prevLease so
> ```

This comment got a little muddled ("an right lease", and there's no longer a `prevLease` argument). 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17109#-:-KpVcAvlNPoW6busxwcP:b-nsz1gj)*
<!-- Sent from Reviewable.io -->
